### ISSUE TITLE
[$2000] Large image doesn't zoom well in Android and gets blurry

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#12893]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:

1. Go to any chat and upload a very long/large image. [Example image](https://user-images.githubusercontent.com/44479856/132889916-e46ac3dc-19ad-4cca-a033-ec13b8ac79ad.jpg)
2. In the Attachment preview, tap the image to zoom
3. Keep zooming until you'll see that the image is blurry

## Expected Result: 
Image should zoom to show a clear image instead of a blurry one

## Actual Result: 
Image content is not visible, and it is blurry.

## Workaround:
unknown

## Platform:

Where is this issue occurring?

- Android

**Version Number:** 1.2.29-6
**Reproducible in staging?:** Y
**Reproducible in production?:** Y
**Email or phone of affected tester (no customers):** 
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** 

https://user-images.githubusercontent.com/43996225/203113972-1f4bb867-186f-4d5c-89e7-cd2761fa2c6b.mp4

https://user-images.githubusercontent.com/43996225/203113166-fb40d7b0-6bb3-418c-ad80-6bb22280e9ba.mov

**Expensify/Expensify Issue URL:** 
**Issue reported by:** @mananjadhav 
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1668928921788569

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~0100609c253c2769f4</li>
        <li>Upwork Job ID: 1597637165026803712</li>
        <li>Last Price Increase: 2022-12-08</li>
    </ul>
</details>

**Comment by aldo-expensify** [comment]
I remember this same images was causing this a long time ago: https://github.com/Expensify/App/issues/5193

I wonder if it ever got fixed and if this is a regression.

**Comment by mananjadhav** [comment]
@aldo-expensify I mentioned this in the slack too, it is a regression. I had tested this earlier and it did work at some point when working on another issue.

**Comment by tjferriss** [comment]
I don't have an Android device. How would you suggest reproducing the bug?

**Comment by mananjadhav** [comment]
@tjferriss do you have an Android emulator or say Browserstack login? You can then reproduce this. This occurs only on Android. Second video on the issue body is from my Android device.

**Comment by aldo-expensify** [comment]
Maybe this will get fixed once the PR https://github.com/Expensify/App/pull/12648 gets re-done and we use fast-image.. maybe

**Comment by mananjadhav** [comment]
Well I tested this when #12648 was worked upon and it wasn't resolved. In fact I found this issue, while testing the same PR. We can still put this on hold, and come back again if you think that's the best option.

**Comment by aldo-expensify** [comment]
> come back again if you think that's the best option.

No, given that you already saw that this doesn't get fixed, ignore my previous comment :P

**Comment by aldo-expensify** [comment]
Reproduced:

<table>
<tr>
<td>
Web
</td>
<td>
Android
</td>
</tr>
<tr>
<td>
<img width="1353" alt="Screen Shot 2022-11-28 at 2 52 43 PM" src="https://user-images.githubusercontent.com/87341702/204397919-77e89891-6b4d-4922-9ada-42a537a13b14.png">
</td>
<td>
<img src="https://user-images.githubusercontent.com/87341702/204398004-7220b1a8-63a0-45c9-b77b-b2254b619691.png">
</td>
</tr>
</table>

**Comment by tjferriss** [comment]
Thank you @aldo-expensify. This looks ready to go so I'm going to add the External tag.

**Comment by aldo-expensify** [comment]
Thanks @markarmanus 

I was also having a look, what you said matches with explanation on what was happening when it was fixed in the past: https://github.com/Expensify/App/issues/5193#issuecomment-974397005

Here is the PR that was done to fix it: https://github.com/Expensify/App/pull/6442

Haven't looked further into why now it is not working, but leaving the links here in case they are of use to someone.

**Comment by markarmanus** [comment]
Proposal

Problem:

The issue is happening on android, when you render an image that the original resolution is massive but you render it with a smaller resolution

Example:

```JSX
// original image is 14500x3000
<Image style={{width: 300, height: 300, }} resize="contain" />
```

On IOS this image is going to stay at the original resolution when zooming in. 

On Android the image gets downscaled to the resolution sepefied to the component so that results in a blury picture on zoom in.

As i stated above this is a recurring issue in the react native community. There is currently no solution for the issue.

I was able to create a solution. its somewhat hacky but it works.
When rendering the image we render the image in its full resolution. then create a container of the image that scales it down to fit the screen.

Here is how to achieve this.
```JS
const imageTransformScale = Math.min(containerHeight / imageHeight, containerWidth / imageWidth);
this.setState({imageHeight, imageWidth, imageTransformScale});
// calculate a imageTransformScale

```

Then Contain the image in a view that uses that scale 

```JSX
<View style={{transform: [{scale: this.state.imageTransformScale}]}}>
     <ImageZoom
         ...
     </ImageZoom>
</View>
```
We also need to extend ImageZoom with these props

```JSX

cropWidth={this.props.windowWidth / this.state.imageTransformScale}
cropHeight={this.state.containerHeight / this.state.imageTransformScale}
```

Here is the result in screen shots.

![image](https://user-images.githubusercontent.com/33489215/204881182-9dc575b2-9977-4f4c-adbf-223cb567cdc6.png)
![image](https://user-images.githubusercontent.com/33489215/204881212-a07a465a-9486-4794-98c7-4c41b72243e6.png)
![image](https://user-images.githubusercontent.com/33489215/204881265-5dc51655-7ca1-4d85-b017-909b325b75bc.png)


Behaves exactly the same on ios

**Comment by tjferriss** [comment]
@marcaaron and @mananjadhav what do we think of @markarmanus proposal?

**Comment by mananjadhav** [comment]
I am not 100% convinced with @markarmanus's proposal. 

> When rendering the image we render the image in its full resolution. then create a container of the image that scales it down to fit the screen.

Also I didn't understand this statement. @markarmanus if you check this earlier, it was solved and it worked fine. I think with multiple refactors to the image component, we've lost/remove the change.

**Comment by markarmanus** [comment]
@mananjadhav If you notice in the last PR we had to fix this issue
https://github.com/Expensify/App/pull/6442

In the code 
It has this in the calculate image size (resolution) 
![image](https://user-images.githubusercontent.com/33489215/205062465-9221b7ef-9067-49ca-8bfb-0865de38eb19.png)

Notice how it does not affect much of the height of the image. Which is why it redners at a really high resolution as in the picture he has in the PR 

![image](https://user-images.githubusercontent.com/33489215/205063006-1a5564f4-2d9b-4485-8490-7e6d7ba8d1dc.png)
But if you notice the picture renders zoomed in already. That because when you pass a big resolution the component it tries to resize the image to a fit a 6000 pixel heigh window. and the phone is obv not 6000 so you see a very zoomed in picture.
if you now go into the code base and manually change the resolution in line 161 in ImageView/index.native.js this will happen.

looking back at the history of this i cant seem to find any time where we had a solution that does not have a high resolution render. 

Also if you take a look at the links i provide above it shows other people having the same issue.

I hope that helps !

**Comment by marcaaron** [comment]
@markarmanus thanks for the investigation, but I think we moved into the "workaround" proposal a bit prematurely.

I'm a little concerned about these comments:

>As i stated above this is a recurring issue in the react native community. There is currently no solution for the issue.
>I was able to create a solution. its somewhat hacky but it works.

Our general approach to fixing bugs is to address issues at their **root cause**. Workarounds are sometimes accepted (usually not) but never without a clear explanation of the problem.

So at a high level, it's possible we might accept this solution (I haven't reviewed it) though unlikely given the lack of explanation into the root cause and focus on the workaround. Let's maybe pause until we have a clear explanation of why it happens. If it's a problem with react native then we should be starting with any open community issues, links to problem code, etc. I see you've added some links [here](https://github.com/Expensify/App/issues/12893#issuecomment-1331563343) - but I think it would be best to summarize what the _**exact problem**_ is before proceeding and make sure everyone is on the same page.

**Comment by mysamark** [comment]
@marcaaron I cant seem to dig up the root cause of the issue. But if you create a react native project from scratch and just use an Image tag with the image above (or any massive image) the same problem happens on android but not ios.

**Comment by markarmanus** [comment]
@marcaaron I cant seem to dig out the cause reason for it. but its a common issue on android. If you create a react native project from scratch. Render the image above and then zoom in it will have the same problem.

A good solution would be to use a Webview component. and we get then git rid of the zoom library complelty which is out dated

**Comment by markarmanus** [comment]
Let me know if you want me to demo Solution 2 or have any questions on it.

**Comment by marcaaron** [comment]
Thanks. No need to demo anything yet. I need to take this to the team and discuss in Slack before we can proceed further. @mananjadhav since you're the assigned C+ do you want to take a stab at summarizing the proposals?

**Comment by marcaaron** [comment]
@mananjadhav what do you think about my previous comment? Are you able to summarize this so we can take it to Slack?

**Comment by mananjadhav** [comment]
@marcaaron I'll summarize this and post it on Slack today.

**Comment by tjferriss** [comment]
@mananjadhav did you start a discussion in Slack today? I don't see anything. Can you tag everyone there when you do so we can follow along?

**Comment by mananjadhav** [comment]
Apologies for the delay. A bit under the weather. I am going to post it in a few hours.

**Comment by aswin-s** [comment]
This bug is caused due to an incorrect image size calculation on 

https://github.com/Expensify/App/blame/main/src/components/ImageView/index.native.js#L75

```jsx
const aspectRatio = Math.min(containerHeight / imageHeight, containerWidth / imageWidth);

imageHeight *= aspectRatio;
imageWidth *= aspectRatio;
```

For the test image above (875px x 14894px ) above lines will max out the height of the image at window height, as it uses `Math.min`. Resizing an extra long image like this to window height will obviously reduce the image detail, resulting in a blurred out view. I think above modification was made to fix an issue where smaller images were not getting fit to window width. ( Related issues https://github.com/Expensify/App/pull/8804, https://github.com/Expensify/App/pull/8285)

That fix seems to be wrong as window width and height shouldn't influence the calculation of the native image dimensions. If we really want to scale the image preview to fit the window bounds, it should be performed via ImageZoom component which has a `centerOn` prop.

My proposal is to revert the changes to `calculateImageSize` so that it is not dependent on windowSize, but still limit the dimensions to be within maxDimensions.

```tsx
calculateImageSize() {
        if (!this.props.url) {
            return;
        }
        ImageSize.getSize(this.props.url).then(({width, height}) => {
            let imageWidth = width;
            let imageHeight = height;
            const containerWidth = Math.round(this.props.windowWidth);
            const containerHeight = Math.round(this.state.containerHeight);

            // Resize the image to max dimensions possible on the Native platforms to prevent crashes on Android. To keep the same behavior, apply to IOS as well.
            const maxDimensionsScale = 11;
            const maxWidth = this.props.windowWidth * maxDimensionsScale;
            const maxHeight = this.props.windowHeight * maxDimensionsScale;
            if (imageWidth > maxWidth || imageHeight > maxHeight) {
                const scale = Math.min(maxWidth / imageWidth, maxHeight / imageHeight);
                imageWidth *= scale;
                imageHeight *= scale;
            }

            this.setState({imageHeight, imageWidth});
        });
    }
```
Then use `centerOn` prop in ImageZoom component to scale and position the original image so that it fits the window on initial load.

Also note that we had switched from Fresco to Glide in #5193, so issue is not related to Fresco downsampling.

**Comment by mananjadhav** [comment]
Started a thread [here](https://expensify.slack.com/archives/C01GTK53T8Q/p1670596654583519)

**Comment by tjferriss** [comment]
@mananjadhav will be summarizing the conclusions from the [Slack discussion](https://expensify.slack.com/archives/C01GTK53T8Q/p1670862866054039?thread_ts=1670596654.583519&cid=C01GTK53T8Q) tomorrow.

**Comment by cristipaval** [comment]
Do we really want to handle very big images? What is the benefit? It is well known that Whatsapp resizes the images and makes them smaller. They don't want to handle big images, there are other original sized photo transfer services like WeTransfer, Airdrop etc. cc @JmillsExpensify for this

If we still want to support very big images, I think we should wait for [this PR](https://github.com/Expensify/App/pull/13304) to get merged and then try to solve this issue. I know you @mananjadhav tested already with `fast-image` integrated (as you mentioned [here](https://github.com/Expensify/App/issues/12893#issuecomment-1329623675)). But I wouldn't dismiss `Glide` wrapped up by `fast-image` yet. I would put this issue on hold until `fast-image` is reintegrated, and then try solving this issue by leveraging `Glide` library with different configs.

**Comment by marcaaron** [comment]
Gonna take this one internal. I think the complexity is fairly high and does not feel like a super urgent bug as it is only affecting very large images (unless I am mistaken on that). There are a lot of good ideas floating around, but I think we need to step back and decide what kind of solution we want.

**Comment by JmillsExpensify** [comment]
Circling back to this issue, I think we should decide what the internal next steps are, as this `bug` just hit the 4 weeks outstanding mark. Here's my two cents: Given that this bug only seems to affect large images, and related, we have several updates in the pipeline for image handling in general, I favor putting a hold on this until we're through @Beamanator's image improvements. Those are much higher value anyway. At that point we can circle back to this consideration and decide what the best next step is. Thoughts?

**Comment by marcaaron** [comment]
That seems reasonable to me. I don't think we have a clear idea of how to proceed here at the moment.

**Comment by JmillsExpensify** [comment]
Filed this issue in the N7 project, along with the larger image improvement issues. More context in [this Slack thread](https://expensify.slack.com/archives/C01GTK53T8Q/p1672088203968479).

**Comment by cristipaval** [comment]
Not overdue, Melvin.

**Comment by marcaaron** [comment]
Still on HOLD

**Comment by tjferriss** [comment]
Still on HOLD. Should this go to `monthly` at this point or do we expect it to come off hold in a few weeks?

**Comment by JmillsExpensify** [comment]
The core image improvements are starting to wind down. I would keep it weekly for a couple more weeks before changing to monthly.

**Comment by marcaaron** [comment]
Seems like this one is still on HOLD. Though - once off HOLD, I think it would be good to do a quick sync on the root cause. I haven't looked at this in a while and still lacking generally clarity on what the solution could be.

**Comment by tjferriss** [comment]
We're still on hold. @JmillsExpensify what exactly should I be tracking to indicate this can be taken off hold?

**Comment by JmillsExpensify** [comment]
Good question. I would think we might even be able to take this off hold now? cc @Beamanator @trjExpensify

**Comment by trjExpensify** [comment]
I think we were specifically holding on image caching, and the mobile side of that is done, it's only web that remains. This is an Android issue, so I think it can come off-hold. Agree with @marcaaron [sentiment here](https://github.com/Expensify/App/issues/12893#issuecomment-1409218632) as a next step.

**Comment by tjferriss** [comment]
Thanks for the context. I've taken the issue off hold. @marcaaron could you kick of this sync on the root causes? Perhaps we can jump back into the [original thread](https://expensify.slack.com/archives/C049HHMV9SM/p1668928921788569) for this discussion.

**Comment by marcaaron** [comment]
@cristipaval are you interested in taking this one? I have a few things on my plate atm and see you are currently assigned. Let me know!

**Comment by cristipaval** [comment]
I'll spend some time on it at the end of this week.

**Comment by marcaaron** [comment]
Cool sounds good! I'm going OOO so un-assigning, but can help with anything when I return. 🙇

**Comment by fabioh8010** [comment]
Hi, I'm Fábio from Callstack and I would like to take a look at this issue.

**Comment by tjferriss** [comment]
Not overdue. @fabioh8010 has been assigned.

**Comment by ArekChr** [comment]
Hey, since @fabioh8010 is working on other tasks, we decided internally that I will take over this task

**Comment by ArekChr** [comment]
### Problem: 
Users on the Android chat app are experiencing a loss of image quality when uploading long or large images and zooming in on them.

### Root Cause: 
The 'react-native-fast-image' framework is used for image rendering and downscales long or large images, leading to image quality issues.

### Solution:
Disable the image transform in the 'react-native-fast-image' framework to prevent image quality loss while rendering in lower resolution than the oryginal size.
Reference: [automatic-transformations-for-imageviews](https://bumptech.github.io/glide/doc/transformations.html#automatic-transformations-for-imageviews)
I can expose the `dontTransform` property in the 'FastImage' component to pass that flag for previews only: `<FastImage dontTransform={true} />`.
To do that, I can create a fork of 'react-native-fast-image' or patch it using patch-package.
Displaying images larger than 8k pixels will also require enabling hardware acceleration and a large heap.

### Alternative Solutions:
Migrate from 'react-native-fast-image' to 'expo-image'. This solution may provide more reliable and optimized image rendering and transformation. 
An example 11k pixels image has been tested using 'expo-images', which displays without downscaling issues while zooming it, the same as on iOS.
Using expo packages in the project will also require installing uni-modules: [installing-expo-modules](https://docs.expo.dev/bare/installing-expo-modules/)
A good reason for choosing this library is that it has a larger community actively working on the library and image performance.

Conclusion: 
To improve the user experience for image previews on the Android app, I will implement one of the proposed solutions to address image quality issues. 
In the first solution, we also need to test displaying images with sizes over 8k pixels on various devices to ensure that enabling hardware acceleration and the large heap will work on low-end devices.

**Comment by cristipaval** [comment]
@ArekChr Lets go with [this](https://expensify.slack.com/archives/C01GTK53T8Q/p1678098494778559?thread_ts=1670596654.583519&cid=C01GTK53T8Q) approach suggested on Slack.

[Here](https://github.com/Expensify/react-native-fast-image) is our current for of the library. Let me know if you have permissions to update it and then to create a pull request with your fix.

**Comment by ArekChr** [comment]
Hey @cristipaval, I will start working on implementing this feature in Expensify/react-native-fast-image fork today 🚀

**Comment by cristipaval** [comment]
Not overdue, @ArekChr is going to open an App PR soon.

**Comment by cristipaval** [comment]
We have our fork ready, we now have to add it to the App.

**Comment by mananjadhav** [comment]
I should be free by tomorrow or day after. Do let me know if I can help in anyway for building the App with the fork.

**Comment by cristipaval** [comment]
We just got past that thing. PR is going to be ready soon, @ArekChr is taking screen recordings now.

**Comment by ArekChr** [comment]
There is a blocker to finish QA checklist tests for mWeb safari and chrome https://github.com/Expensify/App/issues/2499

**Comment by MelvinBot** [comment]
This issue has not been updated in over ***15*** days. @mananjadhav, @tjferriss, @cristipaval, @ArekChr eroding to ***Monthly*** issue.

P.S. Is everyone reading this sure this is really a near-term priority? Be brave: if you disagree, go ahead and close it out. If someone disagrees, they'll reopen it, and if they don't: one less thing to do!

**Comment by cristipaval** [comment]
Closing this one, as [the fix](https://github.com/Expensify/App/pull/16532) hit production a few weeks ago.

**Comment by mananjadhav** [comment]
@cristipaval @tjferriss I did the C+ review and testing on this one. Payout is pending for that .

**Comment by cristipaval** [comment]
> @cristipaval @tjferriss I did the C+ review and testing on this one. Payout is pending for that .

Oh sorry, I missed that! @tjferriss , should we re-open this one and add AwaitingPayment label?

**Comment by tjferriss** [comment]
@mananjadhav I've sent you a request via Upworks here: https://www.upwork.com/jobs/~01ae27c2191acb5934. I also included the reporting bonus in the amount.

**Comment by mananjadhav** [comment]
Accepted @tjferriss

**Comment by tjferriss** [comment]
The payment has been sent.
